### PR TITLE
PPO algorithm which behaves in the same way as baselines.ppo2

### PR DESCRIPTION
**Pull request by emailweixu**
Also added an option for clip_by_global_norm. This PR is on top of PR #58

### CODE REVIEW

File: alf/algorithms/ppo_loss.py [hunk 1]
```diff
@@ -163,3 +170,60 @@ def _pg_loss(self, training_info: TrainingInfo, advantages):
                 policy_gradient_loss, 'policy_gradient_loss')
 
         return policy_gradient_loss
+
+
+@gin.configurable
+class PPOLoss2(PPOLoss):
+    """Create a PPOLoss2 object.
+
+    Note: Different from PPOLoss, PPOLoss uses pre-computed advantage so the
```
Comment (Jialn) [inline alf/algorithms/ppo_loss.py]: PPOLoss2 uses pre-computed advantage so ...

**Comment by emailweixu** [comment]
Merged with the latest master

File: alf/algorithms/ppo_loss.py [hunk 2]
```diff
@@ -52,9 +54,9 @@ def __init__(self,
          + td_loss_weight * td_loss (L^{VF} in equation (9))
          - entropy_regularization * entropy)
 
-        Note: There is a difference with baseline.ppo2 implementation. Here the
+        Note: There is a difference with baselines.ppo2 implementation. Here the
```
Comment (hnyu) [inline alf/algorithms/ppo_loss.py]: Do we still want to keep this PPO loss? It might be misleading to people that this corresponds to PPO1.

File: alf/algorithms/ppo_loss.py [hunk 3]
```diff
@@ -64,10 +66,12 @@ def __init__(self,
                 Q values and returns the loss for each element of the batch.
             use_gae (bool): If True, uses generalized advantage estimation for
                 computing per-timestep advantage. Else, just subtracts value
-                predictions from empirical return.
+                predictions from empirical return. This is ignored by PPOLoss2
+                (Alwasy assumed to be True)
```
Comment (hnyu) [inline alf/algorithms/ppo_loss.py]: Maybe add one more sentence describing the reason of ignoring, e.g., because the advantages are computed in another place using the old policy.

File: alf/examples/ppo2.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+"""
+
+from absl import app
+from absl import flags
+from absl import logging
+
+import gin
+import tensorflow as tf
+
+from alf.examples.off_policy_actor_critic import create_algorithm, create_environment
```
Comment (hnyu) [inline alf/examples/ppo2.py]: No off_policy_actor_critic.py under examples

File: alf/algorithms/ppo_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    It works with ppo_loss.PPOLoss2. It should have same behavior as
+    baselines.ppo2.
+    """
+
+    def __init__(self, algorithm: ActorCriticAlgorithm):
+        assert isinstance(algorithm, ActorCriticAlgorithm)
+        super().__init__(algorithm)
+
+    def preprocess_experience(self, exp: Experience):
+        """Compute advantages and put it into exp.info."""
```
Comment (hnyu) [inline alf/algorithms/ppo_algorithm.py]: Do reward_shaping (if required) before computing the advantages.

File: alf/examples/ppo2_cart_pole.gin [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+create_environment.num_parallel_environments=8
+
+# algorithm config
+PPOLoss2.entropy_regularization=1e-4
+PPOLoss2.gamma=0.98
+PPOLoss2.td_error_loss_fn=@element_wise_huber_loss
+PPOLoss2.normalize_advantages=False
+create_ac_algorithm.actor_fc_layers=(100,)
+create_ac_algorithm.value_fc_layers=(100,)
+create_ac_algorithm.learning_rate=0.001
```
Comment (hnyu) [inline alf/examples/ppo2_cart_pole.gin]: Add rewarding shaping makes the training much more stable and faster:
`ActorCriticAlgorithm.reward_shaping_fn=@reward_scaling
  common.reward_scaling.scale=0.01`

File: alf/algorithms/ppo_algorithm.py (hunk 5 above)
Comment (emailweixu) [reply alf/algorithms/ppo_algorithm.py]: Perhaps we should move reward_shaping_fn to driver or environment wrapper?

File: alf/examples/ppo2.py (hunk 4 above)
Comment (emailweixu) [reply alf/examples/ppo2.py]: This file should be removed.

File: alf/algorithms/ppo_algorithm.py (hunk 5 above)
Comment (emailweixu) [reply alf/algorithms/ppo_algorithm.py]: Done

File: alf/algorithms/ppo_loss.py (hunk 3 above)
Comment (emailweixu) [reply alf/algorithms/ppo_loss.py]: use_gae removed

File: alf/algorithms/ppo_loss.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/ppo_loss.py]: removed

File: alf/algorithms/ppo_algorithm.py [hunk 7]
```diff
@@ -43,6 +43,9 @@ def __init__(self, algorithm: ActorCriticAlgorithm):
 
     def preprocess_experience(self, exp: Experience):
         """Compute advantages and put it into exp.info."""
+        reward = exp.reward
+        if self._algorithm._reward_shaping_fn is not None:
+            reward = self._algorithm._reward_shaping_fn(reward)
         reward = self._algorithm.calc_training_reward(exp.reward, exp.info)
```
Comment (witwolf) [inline alf/algorithms/ppo_algorithm.py]: L49 should be  `reward = self._algorithm.calc_training_reward(reward, exp.info)`

Comment (emailweixu) [reply alf/algorithms/ppo_algorithm.py]: Fixed

File: alf/algorithms/ppo_algorithm.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    def predict(self, time_step: ActionTimeStep, state=None):
+        return self._algorithm.train_step(time_step, state)
+
+    @property
+    def predict_state_spec(self):
+        return self._algorithm.train_state_spec
+
+
+@gin.configurable
+def create_ppo_algorithm(env, debug_summaries=False):
```
Comment (hnyu) [inline alf/algorithms/ppo_algorithm.py]: Can we use a unified create_ac_algorithm interface from from actor_critic_algorithm.py by passing different off_policy_adapter?

File: alf/algorithms/ppo_loss.py [hunk 9]
```diff
@@ -89,9 +85,9 @@ def __init__(self,
             action_spec=action_spec,
             gamma=gamma,
             td_error_loss_fn=td_error_loss_fn,
-            use_gae=use_gae,
+            use_gae=True,
             td_lambda=td_lambda,
-            use_td_lambda_return=use_td_lambda_return,
+            use_td_lambda_return=True,
```
Comment (hnyu) [inline alf/algorithms/ppo_loss.py]: Potentially we can also disable gae or td_lambda for off-policy advantage estimation. Maybe leave this to future.

File: alf/drivers/off_policy_driver.py [hunk 10]
```diff
@@ -232,6 +241,15 @@ def train(self,
...
                     experience)
+                # Make the shape explicit. The shapes of tensors from the
+                # previous line depend on tensor `b`, which is replaced with
+                # None by tf. This makes some operations depending on the shape
+                # of tensor fail. (Currently, it's alf.common.tensor_extend)
+                # TODO: Find a way to work around with shapes containing None
+                # at common.tensor_extend()
+                batch = tf.nest.map_structure(
+                    lambda x: tf.reshape(x, [mini_batch_size] + list(x.shape)[
+                        1:]), batch)
```
Comment (hnyu) [inline alf/drivers/off_policy_driver.py]: Note: in another approved PR this has been changed to a Python loop. So make sure to make them consistent.
